File: app/src/main/java/io/github/shadow578/music_dl/ui/tracks/TracksAdapter.java
package io.github.shadow578.music_dl.ui.tracks;

import android.net.Uri;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.DrawableRes;
import androidx.annotation.NonNull;
import androidx.lifecycle.LifecycleOwner;
import androidx.lifecycle.LiveData;
import androidx.recyclerview.widget.RecyclerView;

import com.bumptech.glide.Glide;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Optional;

import io.github.shadow578.music_dl.KtPorted;
import io.github.shadow578.music_dl.R;
import io.github.shadow578.music_dl.databinding.RecyclerTrackViewBinding;
import io.github.shadow578.music_dl.db.model.TrackInfo;
import io.github.shadow578.music_dl.db.model.TrackStatus;
import io.github.shadow578.music_dl.util.Async;
import io.github.shadow578.music_dl.util.Util;
import io.github.shadow578.music_dl.util.storage.StorageHelper;

/**
 * recyclerview adapter for tracks livedata
 */
@KtPorted
public class TracksAdapter extends RecyclerView.Adapter<TracksAdapter.Holder> {

    @NonNull
    private List<TrackInfo> tracks = new ArrayList<>();

    @NonNull
    private final ItemListener clickListener;

    @NonNull
    private final ItemListener reDownloadListener;

    /**
     * items that should be removed later.
     * key is item position, value if remove was aborted
     */
    @NonNull
    private final HashMap<Integer, Boolean> itemsToDelete = new HashMap<>();

    public TracksAdapter(@NonNull LifecycleOwner owner, @NonNull LiveData<List<TrackInfo>> tracks,
                         @NonNull ItemListener clickListener,
                         @NonNull ItemListener reDownloadListener) {
        this.clickListener = clickListener;
        this.reDownloadListener = reDownloadListener;
        tracks.observe(owner, trackInfoList -> {
            if (trackInfoList == null)
                return;

            this.tracks = trackInfoList;
            notifyDataSetChanged();
        });
    }

    @NonNull
    @Override
    public Holder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        return new Holder(RecyclerTrackViewBinding.inflate(LayoutInflater.from(parent.getContext()), parent, false));
    }

    @Override
    public void onBindViewHolder(@NonNull Holder holder, int position) {
        final TrackInfo track = tracks.get(position);

        // cover
        final Optional<Uri> coverUri = StorageHelper.decodeUri(track.coverKey);
        if (coverUri.isPresent()) {
            // load cover from fs using glide
            Glide.with(holder.b.coverArt)
                    .load(coverUri.get())
                    .placeholder(R.drawable.ic_splash_foreground)
                    .fallback(R.drawable.ic_splash_foreground)
                    .into(holder.b.coverArt);
        } else {
            // load fallback image
            Glide.with(holder.b.coverArt)
                    .load(R.drawable.ic_splash_foreground)
                    .into(holder.b.coverArt);
        }

        // title
        holder.b.title.setText(track.title);

        // build and set artist + album
        final String albumAndArtist;
        if (track.artist != null && track.albumName != null) {
            albumAndArtist = holder.b.getRoot().getContext().getString(R.string.tracks_artist_and_album,
                    track.artist, track.albumName);
        } else if (track.artist != null) {
            albumAndArtist = track.artist;
        } else if (track.albumName != null) {
            albumAndArtist = track.albumName;
        } else {
            albumAndArtist = "";
        }

        holder.b.albumAndArtist.setText(albumAndArtist);

        // status icon
        @DrawableRes final int statusDrawable;
        switch (track.status) {
            case DownloadPending:
                statusDrawable = R.drawable.ic_round_timer_24;
                break;
            case Downloading:
                statusDrawable = R.drawable.ic_downloading_black_24dp;
                break;
            case Downloaded:
                statusDrawable = R.drawable.ic_round_check_circle_outline_24;
                break;
            case DownloadFailed:
                statusDrawable = R.drawable.ic_round_error_outline_24;
                break;
            case FileDeleted:
                statusDrawable = R.drawable.ic_round_remove_circle_outline_24;
                break;
            default:
                throw new IllegalArgumentException("invalid track status: " + track.status);
        }
        holder.b.statusIcon.setImageResource(statusDrawable);

        // retry download button
        final boolean canRetry = track.status.equals(TrackStatus.DownloadFailed) || track.status.equals(TrackStatus.FileDeleted);
        holder.b.retryDownloadContainer.setVisibility(canRetry ? View.VISIBLE : View.GONE);
        holder.b.retryDownloadContainer.setOnClickListener(v -> reDownloadListener.onClick(track));

        // hide on- cover views if retry is shown
        holder.b.statusIcon.setVisibility(canRetry ? View.GONE : View.VISIBLE);
        holder.b.duration.setVisibility(canRetry ? View.GONE : View.VISIBLE);

        // duration
        if (track.duration != null) {
            holder.b.duration.setText(Util.secondsToTimeString(track.duration));
            holder.b.duration.setVisibility(View.VISIBLE);
        } else {
            holder.b.duration.setVisibility(View.GONE);
        }

        // set click listener
        holder.b.getRoot().setOnClickListener(v -> clickListener.onClick(track));


        // deleted mode:
        // setup delete listener
        holder.b.undo.setOnClickListener(v -> {
            itemsToDelete.put(position, false);
            notifyItemChanged(position);
        });

        // make view go into delete mode
        final Boolean isToDelete = itemsToDelete.get(position);
        holder.setDeletedMode(isToDelete != null && isToDelete);
    }

    /**
     * show a undo button for a while, then remove the item
     *
     * @param item           the item to remove
     * @param deleteCallback callback to actually delete the item. called on main thread
     */
    public void deleteLater(@NonNull Holder item, @NonNull ItemListener deleteCallback) {
        final int position = item.getAdapterPosition();
        final TrackInfo track = tracks.get(position);

        // mark as to delete
        itemsToDelete.put(position, true);

        // delete after a delay
        Async.runLaterOnMain(() -> {
            if (Boolean.FALSE.equals(itemsToDelete.get(position))) {
                return;
            }

            // remove from map
            itemsToDelete.remove(position);

            // animate removal nicely
            notifyItemRemoved(position);

            // actually remove
            Async.runLaterOnMain(() -> deleteCallback.onClick(track), 100);
        }, 5000);

        // update to reflect new deleted state
        notifyItemChanged(position);
    }

    @Override
    public int getItemCount() {
        return tracks.size();
    }

    /**
     * a view holder for the items of this adapter
     */
    public static class Holder extends RecyclerView.ViewHolder {
        /**
         * view binding of the view this holder holds
         */
        public final RecyclerTrackViewBinding b;

        public Holder(@NonNull RecyclerTrackViewBinding b) {
            super(b.getRoot());
            this.b = b;
        }

        /**
         * set if the 'deleted mode' view should be used
         *
         * @param deletedMode is this item in deleted mode?
         */
        public void setDeletedMode(boolean deletedMode) {
            b.containerMain.setVisibility(deletedMode ? View.INVISIBLE : View.VISIBLE);
            b.containerUndo.setVisibility(deletedMode ? View.VISIBLE : View.GONE);
        }
    }

    /**
     * a click listener for track items
     */
    @FunctionalInterface
    public interface ItemListener {
        /**
         * called when a track view is selected
         *
         * @param track the track the view shows
         */
        void onClick(@NonNull TrackInfo track);
    }
}

File: app/src/main/kotlin/io/github/shadow578/yodel/ui/tracks/TracksAdapter.kt
package io.github.shadow578.yodel.ui.tracks

import android.view.*
import androidx.annotation.DrawableRes
import androidx.lifecycle.LifecycleOwner
import androidx.lifecycle.LiveData
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.RecyclerView
import com.bumptech.glide.Glide
import io.github.shadow578.yodel.R
import io.github.shadow578.yodel.databinding.RecyclerTrackViewBinding
import io.github.shadow578.yodel.db.model.*
import io.github.shadow578.yodel.util.*
import io.github.shadow578.yodel.util.storage.decodeToUri
import kotlinx.coroutines.delay
import java.util.*
import kotlin.collections.set

/**
 * recyclerview adapter for tracks livedata
 */
class TracksAdapter(
        owner: LifecycleOwner,
        tracks: LiveData<List<TrackInfo>>,
        private val clickListener: ItemListener,
        private val reDownloadListener: ItemListener
) : RecyclerView.Adapter<TracksAdapter.Holder?>() {
    init {
        tracks.observe(owner, { newTracks: List<TrackInfo> ->
            // calculate difference
            val difference = DiffUtil.calculateDiff(object : DiffUtil.Callback() {
                override fun getOldListSize(): Int {
                    return currentTracks.size
                }

                override fun getNewListSize(): Int {
                    return newTracks.size
                }

                override fun areItemsTheSame(oldItemPosition: Int, newItemPosition: Int): Boolean {
                    return currentTracks[oldItemPosition].id == newTracks[newItemPosition].id
                }

                override fun areContentsTheSame(oldItemPosition: Int, newItemPosition: Int): Boolean {
                    return currentTracks[oldItemPosition].equalsContent(newTracks[newItemPosition])
                }
            })

            // update data and apply the changes
            currentTracks = newTracks
            difference.dispatchUpdatesTo(this)
        })
    }

    /**
     * current tracks list displayed
     */
    private var currentTracks: List<TrackInfo> = listOf()

    /**
     * items that should be removed later.
     * key is item position, value if remove was aborted
     */
    private val itemsToDelete = HashMap<Int, Boolean>()

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): Holder {
        return Holder(
                RecyclerTrackViewBinding.inflate(
                        LayoutInflater.from(parent.context),
                        parent,
                        false
                )
        )
    }

    override fun onBindViewHolder(holder: Holder, position: Int) {
        val track = currentTracks[position]

        // cover
        val coverUri = track.coverKey.decodeToUri()
        if (coverUri != null) {
            // load cover from fs using glide
            Glide.with(holder.b.coverArt)
                    .load(coverUri)
                    .placeholder(R.drawable.ic_splash_foreground)
                    .fallback(R.drawable.ic_splash_foreground)
                    .into(holder.b.coverArt)
        } else {
            // load fallback image
            Glide.with(holder.b.coverArt)
                    .load(R.drawable.ic_splash_foreground)
                    .into(holder.b.coverArt)
        }

        // title
        holder.b.title.text = track.title

        // build and set artist + album
        val albumAndArtist: String? =
                if (track.artist != null && track.albumName != null) {
                    holder.b.root.context.getString(
                            R.string.tracks_artist_and_album,
                            track.artist,
                            track.albumName
                    )
                } else if (track.artist != null) {
                    track.artist
                } else if (track.albumName != null) {
                    track.albumName
                } else {
                    null
                }
        holder.b.albumAndArtist.text = albumAndArtist ?: ""

        // status icon
        @DrawableRes val statusDrawable: Int = when (track.status) {
            TrackStatus.DownloadPending -> R.drawable.ic_round_timer_24
            TrackStatus.Downloading -> R.drawable.ic_downloading_black_24dp
            TrackStatus.Downloaded -> R.drawable.ic_round_check_circle_outline_24
            TrackStatus.DownloadFailed -> R.drawable.ic_round_error_outline_24
            TrackStatus.FileDeleted -> R.drawable.ic_round_remove_circle_outline_24
        }
        holder.b.statusIcon.setImageResource(statusDrawable)

        // retry download button
        val canRetry = track.status == TrackStatus.DownloadFailed
                || track.status == TrackStatus.FileDeleted
        holder.b.retryDownloadContainer.visibility = if (canRetry) View.VISIBLE else View.GONE
        holder.b.retryDownloadContainer.setOnClickListener { reDownloadListener.onClick(track) }

        // hide on- cover views if retry is shown
        holder.b.statusIcon.visibility = if (canRetry) View.GONE else View.VISIBLE
        holder.b.duration.visibility = if (canRetry) View.GONE else View.VISIBLE

        // duration
        if (track.duration != null) {
            holder.b.duration.text = track.duration?.secondsToTimeString()
            holder.b.duration.visibility = View.VISIBLE
        } else {
            holder.b.duration.visibility = View.GONE
        }

        // set click listener
        holder.b.root.setOnClickListener { clickListener.onClick(track) }


        // deleted mode:
        // setup delete listener
        holder.b.undo.setOnClickListener {
            itemsToDelete[position] = false
            notifyItemChanged(position)
        }

        // make view go into delete mode
        val isToDelete = itemsToDelete[position]
        holder.setDeletedMode(isToDelete != null && isToDelete)
    }

    /**
     * show a undo button for a while, then remove the item
     *
     * @param item           the item to remove
     * @param deleteCallback callback to actually delete the item. called on main thread
     */
    fun deleteLater(item: Holder, deleteCallback: ItemListener) {
        val position = item.bindingAdapterPosition
        val track = currentTracks[position]

        // mark as to delete
        itemsToDelete[position] = true

        // delete after a delay
        launchMain {
            delay(5000)
            if (itemsToDelete[position] == false)
                return@launchMain

            // remove from map
            itemsToDelete.remove(position)

            // animate removal nicely
            notifyItemRemoved(position)

            // actually remove after short delay
            delay(100)
            deleteCallback.onClick(track)
        }

        // update to reflect new deleted state
        notifyItemChanged(position)
    }

    override fun getItemCount(): Int {
        return currentTracks.size
    }

    /**
     * a view holder for the items of this adapter
     */
    class Holder(
            /**
             * view binding of the view this holder holds
             */
            val b: RecyclerTrackViewBinding
    ) : RecyclerView.ViewHolder(b.root) {

        /**
         * set if the 'deleted mode' view should be used
         *
         * @param deletedMode is this item in deleted mode?
         */
        fun setDeletedMode(deletedMode: Boolean) {
            b.containerMain.visibility = if (deletedMode) View.INVISIBLE else View.VISIBLE
            b.containerUndo.visibility = if (deletedMode) View.VISIBLE else View.GONE
        }

    }

    /**
     * a click listener for track items
     */
    fun interface ItemListener {
        /**
         * called when a track view is selected
         *
         * @param track the track the view shows
         */
        fun onClick(track: TrackInfo)
    }
}